python/delta/tables.py

```diff
@@ -224,114 +333,206 @@ def _condition_to_jcolumn(cls, condition, argname="'condition'"):
+            })
+
+    - Constraints in the ``whenNotMatched`` clauses:
+
+      - This clause can have only an ``insert`` action, which can have an optional condition.
+
+      - If ``whenNotMatchedInsert`` is not present or if it is present but the non-matching
+        source row does not satisfy the condition, then the source row is not inserted.
+
+      - If you want to insert all the columns of the target Delta table with the
+        corresponding column of the source DataFrame, then you can use
+        ``whenNotMatchedInsertAll()``. This is equivalent to::
+
+            whenMatchedInsert(values = {
+              "col1": "source.col1",
+              "col2": "source.col2",
+              ...    # for all columns in the delta table
+            })
+
+    Example 1 with conditions and update expressions as SQL formatted string::
+
+        deltaTable.alias("events").merge(
+            source = updatesDF.alias("updates"),
+            condition = "events.eventId = updates.eventId"
+          ).whenMatchedUpdate(set =
+            {
+              "data": "updates.data",
+              "count": "events.count + 1"
+            }
+          ).whenNotMatchedInsert(values =
+            {
+              "date": "updates.date",
+              "eventId": "updates.eventId",
+              "data": "updates.data",
+              "count": "1"
+            }
+          ).execute()
+
+    Example 2 with conditions and update expressions as Spark SQL functions::
+
+        from pyspark.sql.functions import *
+
+        deltaTable.alias("events").merge(
+            source = updatesDF.alias("updates"),
+            condition = expr("events.eventId = updates.eventId")
+          ).whenMatchedUpdate(set =
+            {
+              "data" : col("updates.data"),
+              "count": col("events.count") + 1
+            }
+          ).whenNotMatchedInsert(values =
+            {
+              "date": col("updates.date"),
+              "eventId": col("updates.eventId"),
+              "data": col("updates.data"),
+              "count": lit("1")
+            }
+          ).execute()
+
+    .. versionadded:: 0.4
 
     .. note:: Evolving
     """
     def __init__(self, spark, jbuilder):
         self._spark = spark
         self._jbuilder = jbuilder
 
+    @since(0.4)
     def whenMatchedUpdate(self, condition=None, set=None):
         """
         Update a matched table row based on the rules defined by ``set``.
         If a ``condition`` is specified, then it must evaluate to true for the row to be updated.
 
+        See :py:class:`~delta.tables.DeltaMergeBuilder` for complete usage details.
+
+        :param condition: Optional condition of the update
+        :type condition: str or pyspark.sql.Column
+        :param set: Defines the rules of setting the values of columns that need to be updated.
+                    *Note: This param is not Optional.* Default value None is present to allow
```

Comment:
That's... pretty confusing. I'd say required instead of not optional, and I hope we throw a good error for this.

Reply:
we do.